### PR TITLE
Enabled HTTP/2 protocol - fix #9530

### LINKED ISSUE #9530: [Feature Request] Support Multiplexing of Proxied Datasource Queries
Grafana dashboards with high information density (i.e. lots of small panels) can be slow to load since only ~6 panels can be loading at a time. This is because of two main current limitations:

1. Grafana does not allow for a single datasource query to map to multiple panels
2. With HTTP/1.1, browsers only allow [about 6 simultaneous queries to a single domain](https://stackoverflow.com/questions/985431/max-parallel-http-connections-in-a-browser) so we end up with head of line blocking. See [Why is HTTP/2 multiplexed](https://http2.github.io/faq/#why-is-http2-multiplexed)

To get a feel for this, the [Golang http2 Gophertiles demo](https://http2.golang.org/gophertiles) can give a good feeling of loading lots of "tiles", which are analogous to panels, with a higher latency -- what you might get from more expensive time series queries. 

The query-per-panel limitation will also often not fit database performance characteristics for certain types of dense dashboards. Consider the following dashboards from Opserver and trying to recreate them in Grafana:

Host Overview:
screen shot 2017-10-15 at 9 32 07 am

Switch Interfaces:
screen shot 2017-10-15 at 9 35 56 am

In any datasource where for example CPU, Memory, Network are stored together, multiple panels would share a scan/index like query ("get cpu for all hosts") and having say 50-100 individual queries won't perform well.

In particular with the new grid engine (mentioned [here by torkel](https://community.grafana.com/t/table-panel-span-multiple-rows/2945)) I expect people will try to create more of these sorts of dashboards and run into these performance problems with real world latency against TSDBs.

A waterfall Network load from Chrome illustrated this well. The gray bars are time that panels have not actually started, "stalled", because they are waiting for other panels to load first:

screen shot 2017-10-15 at 9 46 30 am

### PR DESCRIPTION
**Pull request by jangaraj**
WIP - feature request: https://github.com/grafana/grafana/issues/9530

Question for discussion:
1.) Do we want to support HTTP/2 protocol?
It will be available only for TLS connections. It may save a few ms for all users. In theory, it may improve performance for dashboards with many panels. Final implementation will require a change of default TLS configuration - some bad ciphers are blocked for HTTP/2 protocol.

Chrome test with 100 graph panels with random data doesn't show any significant improvement:
HTTP/2:
image

HTTP/1.1:
image

### CODE REVIEW

File: pkg/api/http_server.go [hunk 1]
```diff
@@ -143,25 +143,16 @@ func (hs *HTTPServer) listenAndServeTLS(certfile, keyfile string) error {
...
-			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
-			tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
-			tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
-			tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
-			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
-			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
 			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
-			tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+			tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,  // the best cipher for mobile devices without HW AES support - Snapdragon 800-
+			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, // must be enabled for http/2 support
```
Comment (torkelo) [inline pkg/api/http_server.go]: Can you explain the changes here why the change to PreferServerCipherSuites, and the removal of cipher, what implications does this change have?

Comment (jangaraj) [reply pkg/api/http_server.go]: Some weak ciphers are on the h2 blacklist - https://github.com/golang/go/blob/master/src/net/http/h2_bundle.go#L397-L686, so they can't be used.

PreferServerCipherSuites - there is no official recommendation for client/server cipher preference. Generally, if support for weak ciphers is required, then server ciphers should be preferred. If you have a support only for secure ciphers, then the client can decide which cipher is the best one. The best blog post for cipher problem: https://hynek.me/articles/hardening-your-web-servers-ssl-ciphers/

This is WIP and it needs a discussion - for example list and order of supported ciphers.

**Comment by bergquist** [comment]
I don't think we are ready for this. 

Making good use of http2 would currently require the frontend to batch requests. Unless we intent to always require TLS (which we won't. people should be able to install Grafana locally without bothering with TLS) we would have to maintain two versions for how to requests metrics from the panel. 

I think a better step would be to enable batch queries from the frontend and have the backend send multiple requests to the datasource in the backend. Basically having the panel send all its settings to the backend and let it fetch all the data points in one go.

**Comment by marefr** [comment]
Side note: To get some direct benefits of HTTP/2 from browser the most easiest thing is probably to enable HTTP/2 in a reverse proxy.

**Comment by jangaraj** [comment]
Thank you for your feedback. I agree with your point, that user can enable HTTP/2 in a reverse proxy. So my first question is if you want to support HTTP/2 directly in the Grafana. Or you can use the same approach as you use for complex TLS configuration - user can configure TLS in own reverse proxy/load balancer.

Batching of requests in the frontend will help with panel queries. HTTP/2 backend implementation may improve the performance of all requests (static files as well).

If you don't want to support HTTP/2, then please feel free to close this PR.
Otherwise, I will try to solve mentioned problems:
- ciphers - IMHO it will need only better order
- HTTP/2 for non TLS listener - `h2c` implementation https://godoc.org/golang.org/x/net/http2/h2c

**Comment by torkelo** [comment]
I http2 support would be great, but don’t want do break any existing setups

**Comment by torkelo** [comment]
Closing this as there has been no activity here for a while. Think http2 would be good to support natively but some option to disable it might be good if it causes issues (same with the cert changes in this PR).
